packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,449 @@
+      return TextEditingDeltaReplacement(
+        oldText: oldText,
+        deltaText: replacementSource,
+        deltaRange: TextRange(
+          start: replacementDestinationStart,
+          end: replacementDestinationEnd,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+    
+    assert(false);
+    return TextEditingDeltaNonTextUpdate(
+      oldText: oldText,
+      selection: newSelection,
+      composing: newComposing,
+    );
+  }
+
+  /// The old text state before the delta has occurred.
+  final String oldText;
+
+  /// The value represents the text that is being inserted/deleted by this delta.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType]:
+  ///
+  ///  * For a [TextEditingDeltaType.insertion] this will be the character/s being
+  /// inserted.
+  ///
+  ///  * For a [TextEditingDeltaType.deletion] this will be an empty string.
+  ///
+  ///  * For a [TextEditingDeltaType.replacement] this will be the text that is
+  /// replacing the [TextEditingDelta.deltaRange].
+  ///
+  ///  * For a [TextEditingDeltaType.nonTextUpdate] this will be an empty string.
+  final String deltaText;
+
+  /// This value can either represent a range of text that the delta is changing
+  /// or if is a collapsed range then it represents the point where this delta
+  /// began.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType]:
+  ///
+  ///  * For a [TextEditingDeltaType.insertion] this will be a collapsed range
+  /// representing the cursor position where the insertion began.
+  ///
+  ///  * For a [TextEditingDeltaType.deletion] this will be the range of text
+  /// that was deleted.
+  ///
+  ///  * For a [TextEditingDeltaType.replacement] this will be the range of
+  /// characters that are being replaced.
+  ///
+  ///  * For a [TextEditingDeltaType.nonTextUpdate] this will be a collapsed range
+  /// of (-1,-1).
+  final TextRange deltaRange;
+
+  /// {@template flutter.services.TextEditingDelta.deltaType}
+  /// The type of delta that has occured.
+  ///
+  /// See [TextEditingDeltaType] for more information.
+  /// {@endtemplate}
+  TextEditingDeltaType get deltaType;
+
+  /// The range of text that is currently selected after the delta has been
+  /// applied.
+  final TextSelection selection;
+
+  /// The range of text that is still being composed after the delta has been
+  /// applied.
+  final TextRange composing;
+
+  /// {@template flutter.services.TextEditingDelta.apply}
+  /// This method will take the given [TextEditingValue] and return a new
+  /// [TextEditingValue] with that instance of [TextEditingDelta] applied to it.
+  /// {@endtemplate}
+  TextEditingValue apply(TextEditingValue value) {
+    // Verify that the delta we are applying is applicable to the given
+    // editing state. If not then we return the original value.
+    if (value.text != oldText) {
+      return value;
+    }
+
+    return apply(value);
```

Comment:
This `apply()` calls calls the subclass implementation. I left this implemented because all `apply` method's should share the common delta verification check.

Reply:
The `apply` invocation will be directly dispatched to the subclass and the `TextEditingDelta.apply` will be overriden. The super implementation will not be invoked unless you explicitly call `super.apply`.